### PR TITLE
Remove py2 support

### PR DESCRIPTION
**Pull request by ChrisBarker-NOAA**
This PR removes the six dependency, and all other Py2 compatibility code I could find.

Still a lot of `u"strings"`, but they do no harm.

Also a few changes to remove deprecation warnings.

And a small change to make pytest a bit happier -- I really like pytest for running tests.

I haven't run coverage, so not completely sure if it's been fully tested, but I did make sure most of teh changes broke a test before I fixed it, so it's certainly mostly there.

I couldn't test without six installed, as it's needed by a dependency, but I did grep for "six", and I'm pretty sure I got it all.

### CODE REVIEW

**Comment by stankudrow** [comment]
@ChrisBarker-NOAA , hello.

Could you pass all python files via `pyupgrade`? Or maybe there are other solutions?

**Comment by Kludex** [comment]
Don't run any automated tools yet, please. It's easier to review manual changes first.

File: bson/tests/test_object.py [hunk 1]
```diff
@@ -3,8 +3,8 @@
 
 from bson import BSONCoding, dumps, loads, import_class
 
-
-class TestData(BSONCoding):
+# named with underscore so as not to confuse pytest
+class _TestData(BSONCoding):
```
Comment (stankudrow) [inline bson/tests/test_object.py:7]: https://github.com/py-bson/bson/pull/129#discussion_r1889545037

File: bson/tests/test_object.py [hunk 2]
```diff
@@ -31,12 +31,12 @@ def __ne__(self, other):
 
 class TestObjectCoding(TestCase):
     def test_codec(self):
-        import_class(TestData)
-        data = TestData(u"Lorem ipsum dolor sit amet",
+        import_class(_TestData)
+        data = _TestData(u"Lorem ipsum dolor sit amet",
```
Comment (stankudrow) [inline bson/tests/test_object.py:35]: Unicode string and further below.

File: bson/tests/test_objectid.py [hunk 3]
```diff
@@ -511,18 +510,19 @@ def test_pid(self):
         self.assertTrue(oid_generated_on_client(ObjectId()))
 
     def test_generation_time(self):
-        d1 = datetime.datetime.utcnow()
+        d1 = datetime.datetime.now(datetime.UTC)
```
Comment (stankudrow) [inline bson/tests/test_objectid.py:513]: Will it work in Py3.9?

File: bson/tests/test_objectid.py [hunk 4]
```diff
@@ -511,18 +510,19 @@ def test_pid(self):
...
-        d2 = d2.replace(tzinfo=None)
         self.assertTrue(d2 - d1 < datetime.timedelta(seconds=2))
 
     def test_from_datetime(self):
         if 'PyPy 1.8.0' in sys.version:
             # See https://bugs.pypy.org/issue1092
             raise SkipTest("datetime.timedelta is broken in pypy 1.8.0")
-        d = datetime.datetime.utcnow()
+        # note: could use tz aware datetimes here?
+        d = datetime.datetime.now(datetime.UTC).replace(tzinfo=None)
```
Comment (stankudrow) [inline bson/tests/test_objectid.py:524]: Py3.9?

File: bson/tests/test_objectid.py (hunk 3 above)
Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py:513]: Darn -- nope.

Which is really annoying -- why depreciate when supported versions don't support the new way? 

Good reason to get the CI going! 

I don't have time to fix this right now, but I allowed maintainers to push to my branch, so you can do it -- or wait for me.

File: bson/tests/test_objectid.py (hunk 4 above)
Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py:524]: Hmm -- 3.9 is not so far from end of life -- time to dump it now? the last version will keep working .....

Comment (stankudrow) [reply bson/tests/test_objectid.py:524]: https://endoflife.date/python - the security support ends in 10 months, so it is maintained in a way.

File: bson/tests/test_objectid.py (hunk 3 above)
Comment (stankudrow) [reply bson/tests/test_objectid.py:513]: Yep, and a PR for the CI/CD is already opened: https://github.com/py-bson/bson/pull/131

File: bson/tests/test_objectid.py (hunk 4 above)
Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py:524]: > the security support ends in 10 months, so it is maintained in a way.

yes it's not end of life yet -- but will be in 10 months, so is it worth supporting in the next version of this package? your call :-).

File: bson/tests/test_objectid.py (hunk 3 above)
Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py]: OK -- fixed this

File: bson/tests/test_objectid.py (hunk 4 above)
Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py]: turns out there was very little to do to keep it py3.9 compatible, so I've done that.

Comment (ChrisBarker-NOAA) [reply bson/tests/test_objectid.py]: > And do you need to compute now with UTC and then replacing the latter with None?

well, I agree, but that's the new way to do it -- the naive utcnow() is deprecated -- I don't think it should be -- a naive utc datetime is useful, but there you go.

I figure while updating, now is the time to address deprecations.

**Comment by ChrisBarker-NOAA** [comment]
All looks good now -- passing on py 3.9 -- 3.13.

There some more formatting cleanup to be done, but as suggested, I think a separate PR is called for there.

**Comment by ChrisBarker-NOAA** [comment]
Once #129 is merged, we'll probably want to dump the changes made here in the workflow.

**Comment by stankudrow** [comment]
> Once #129 is merged, we'll probably want to dump the changes made here in the workflow.

Already merged with other PRs, so when rebased, I'll look through as soon as possible.

**Comment by stankudrow** [comment]
@ChrisBarker-NOAA , herllo. Are you ready to rebase the PR and we can go on with reviewing?

**Pull request closed without merging**